qiskit/pulse/configuration.py

```diff
@@ -41,6 +58,14 @@ def __repr__(self):
             ", ".join(f"{str(k)}={str(v)}" for k, v in self.params.items()),
         )
 
+    def __eq__(self, other):
+        if isinstance(other, Kernel):
+            return _assert_nested_dict_equal(self.__dict__, other.__dict__)
+        return False
+
+    def __hash__(self):
```

Comment:
Was there a need for the hash? I didn't notice the kernel being used in a set or dictionary.

Reply:
If `Kernel` is unhashable, the following test fails:
https://github.com/Qiskit/qiskit-terra/blob/b73115f80bc575b5dbc8b98ad003cd3b4ff9c00d/test/python/pulse/test_instructions.py#L68-L101

Reply:
Hmm, I am not sure what the best approach is here.

The current behavior is that the kernel and discriminator use Python's default hashing and equality behavior where every instance gets a unique hash and the only equality is that an instance is equal to itself. So they are hashable but not in a particularly useful way. Overriding the default `__eq__` method unsets the default `__hash__` method. I wonder if other behavior relies on pulse instructions being hashable? I have not tried to check.

The kernel and discriminator classes are not well-specified -- basically just json serializable structures (extended to support numpy arrays) -- so nested dicts and lists of float, int, str, and ndarrays. That worked fine for qobj but does not fit as well for qpy. They don't need to be mutable so hashing the data should be okay. I think a more proper hash would be to replace lists with tuples and dicts with tuples of key-value tuples recursively, but that is a bit involved, especially if we don't really need hashing.

Reply:
> I wonder if other behavior relies on pulse instructions being hashable? I have not tried to check.

`SymbolicPulse` is not hashable:
https://github.com/Qiskit/qiskit-terra/blob/e9f8b7c50968501e019d0cb426676ac606eb5a10/qiskit/pulse/library/symbolic_pulses.py#L570

So, not all of instructions are actually hashable.

Reply:
>  I think a more proper hash would be to replace lists with tuples and dicts with tuples of key-value tuples recursively, but that is a bit involved, especially if we don't really need hashing.

Thank you so much for your suggestion!
Although I am still not sure whether `Kernel` and `Discriminator` should remain hashable (because the other operand `SymbolicPulse` is unhashable and as a result not all of instructions are hashable), as long as we cannot eliminate the possibility that their hashing is used somewhere, I suppose that it is safer to keep them hashable. 
I updated the hash methods following your suggestion :) 👉  https://github.com/Qiskit/qiskit-terra/pull/10327/commits/0afd18dde0e3041eb3ae7e40492ea55d11fdb89d